### DISCUSSION TITLE
useWatch,getValues return sometimes undefined for dynamic index update

### DISCUSSION THREAD

**Discussion by jandolezal71** [react-hook-form/react-hook-form#10517] [Q&A]
### Version Number

7.44.3

### Codesandbox/Expo snack

https://codesandbox.io/s/reverent-mclaren-yn4fnr?file=/src/App.js

### Steps to reproduce

1. Go to codesandbox url
2. Click on edit button in last row
2. Click on edit button in row before

---

Use case:

I have table, where customer can expand row by clicking on edit in row. This edit will open next row with hook form, where he can edit row data.  Before table is submit button, which will submit all edited items (only edited, not all rows). I need to map table row indexes exactly to indexes in field array to properly propagate future errors.

When row expand, i will update() data on specific index. When i will update fe. index 2, field array has [undefined, undefined, data]. 

But when i will access this item in nested component with either `getValues('items.${index}.value')` or `useWatch({ control, name: 'items.${index}.value' })`, this value is SOMETIMES undefined and later logic will fail :( 

Next issue is, `useWatch()` dont see full array `[undefined, undefined, data]`, but only `[data]` which will break indexes. `getValues` will see `[undefined, undefined, data]`, but SOMETIMES return undefined instead of data.

If you replace `useWatch` with `getValues` in codesandobx example in nested row, it will work SOMETIMES. If you click some rows, this will laer throw an error anyway.

Would you suggest some other design how to solve this? Or this is a bug?

### Expected behaviour

Every row should get data from specific index in field array

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
React hook form is made for forms, i don't see any inputs. perhaps you are using the wrong tool for what's your working on. useFieldArray create fields object which contain key of each row of inputs and not used too.

**Comment by jandolezal71** [comment]
> 

I know, codesandbox was simplified without inputs. Core point is, there is missing data from these methods. In real world i use inputs too ofc. In every row form, every input is registered to specific index.

Whole idea is working, but problem is, that useWach/getValues sometime dont return data for that index.

I do not see point to create useForm() for every form in row. This way i will create parent useForm for all expanded rows and only submit these.

Just to say, useWatch() is there for displaying some other concatenated data in row, not for inputs itself.

**Reply by jandolezal71** [reply]
![image](https://github.com/react-hook-form/react-hook-form/assets/46811197/b80fec65-4716-4fe6-8276-528f1ad12844)

**Comment by leapful** [comment]
@jandolezal71 

It seems you are mis-understanding how react-hook-form works. You are using `rows` array object to render the `Table` component with row item but then use `useFieldArray` to get the editable value. Since there is no control of react-hook-form on the data, the value will be out of sync and leads to undefined error.

You need to render the row items by using `fields` object from `useFieldArray` hook instead.

Working demo: https://codesandbox.io/s/condescending-brahmagupta-f94r94?file=/src/App.js
